### PR TITLE
Introduce Utility/Functional

### PR DESCRIPTION
**Pull request by jokva**
Utility/Functional is a lightweight high level functional-oriented sub library that attempts to abstract some common uses and boilerplate around the parser code (and later maybe for other modules to use).

This patch introduce only three functions, but they have proven common enough to warrant some common implementation.

I plan for things here to be added on a by-need basis.

### CODE REVIEW

**Comment by joakim-hove** [comment]
Gee - looking at your beautiful documentation even I can see the benefit of documentation.

File: opm/parser/eclipse/Utility/Functional.hpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+     *
+     * int plus1( int x ) { return x + 1; }
+     * base_vec = { 0, 1, 2, 3, 4 };
+     * vec = fun::map( &plus1, base_vec );
+     *
+     * vec => { 1, 2, 3, 4, 5 }
+     *
+     * --
+     *
+     * int mul2 = []( int x ) { return x * 2; };
```
Comment (atgeirr) [inline opm/parser/eclipse/Utility/Functional.hpp]: Minor comment glitch: `int` should be `auto`.

File: opm/parser/eclipse/Utility/Functional.hpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                []( std::size_t acc, const std::vector< A >& x ) {
+                    return acc + x.size();
+                    }
+                );
+
+        std::vector< A > dst;
+        dst.reserve( size );
+
+        for( auto& x : src )
+            std::move( x.begin(), x.end(), std::back_inserter( dst ) );
```
Comment (atgeirr) [inline opm/parser/eclipse/Utility/Functional.hpp]: I cannot find a reference for this way to call move, is it really C++11?

Comment (jokva) [inline opm/parser/eclipse/Utility/Functional.hpp]: Yes. http://en.cppreference.com/w/cpp/algorithm/move

File: opm/parser/eclipse/Utility/Functional.hpp (hunk 1 above)
Comment (jokva) [inline opm/parser/eclipse/Utility/Functional.hpp]: Absolutely so.

File: opm/parser/eclipse/Utility/Functional.hpp (hunk 2 above)
Comment (atgeirr) [inline opm/parser/eclipse/Utility/Functional.hpp]: Thanks, googling for std::move I only found the <utility> version and not the <algorithm> version, a bit strange that the two overloads are in different headers!

Comment (jokva) [inline opm/parser/eclipse/Utility/Functional.hpp]: Nine times out of ten it's the std::move from utility you're looking for, as std::move(begin, end, output) is rather specific to the algorithms library. Confusing maybe, but makes sense in my opinion. It's a *very* useful feature however, as demonstrated here.

**Pull request merged**